Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js

```diff
@@ -0,0 +1,181 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+WI.DeviceSettingsManager = class DeviceSettingsManager extends WI.Object
+{
+    constructor()
+    {
+        super();
+
+        this._overridenDeviceUserAgent = null;
+        this._overridenDeviceScreenSize = null;
+        this._overridenDeviceSettings = new Map;
+    }
+
+    // Target
+
+    initializeTarget(target)
+    {
+        if (target.hasDomain("Page")) {
+            // COMPATIBILITY (iOS 12.2): Page.overrideUserAgent did not exist.
+            if (target.hasCommand("Page.overrideUserAgent") && this._overridenDeviceUserAgent)
+                target.PageAgent.overrideUserAgent(this._overridenDeviceUserAgent);
+
+            // COMPATIBILITY (iOS 12.2): Page.overrideSetting did not exist.
+            if (target.hasCommand("Page.overrideSetting")) {
+                for (let [setting, value] of this._overridenDeviceSettings)
+                    target.PageAgent.overrideSetting(setting, value);
+            }
+        }
+    }
+
+    // Static
+
+    static supportsSetScreenSizeOverride()
+    {
+        return InspectorBackend.hasCommand("Page.setScreenSizeOverride");
+    }
+
+    // Public
+
+    get defaultSettingsChanged() {  return this._overridenDeviceUserAgent || this._overridenDeviceSettings.size > 0; }
+
+    get overridenDeviceUserAgent() { return this._overridenDeviceUserAgent; }
+
+    get overridenDeviceScreenSize() { return this._overridenDeviceScreenSize; }
+
+    get overridenDeviceSettings() { return this._overridenDeviceSettings; }
+
+    overrideDeviceSetting(setting, value, callback)
+    {
+        let target = WI.assumingMainTarget();
+
+        if (!target.hasCommand("Page.overrideSetting"))
+            return;
+
+        if (this._overridenDeviceSettings.has(setting)) {
```

Comment:
you could avoid repeating a bunch of code using `invoke`
```
let commandArguments = {
    setting,
};

let removed = false;
if (! this._overridenDeviceSettings.has(setting)) {
    removed = true;
    commandArguments.value = value;
}

target.PageAgent.overrideSetting.invoke(commandArguments, (error) => {
    if (error) {
        WI.reportInternalError(error);
        return;
    }

    if (removed)
        this._overriddenDeviceSettings.delete(setting);
    else
        this._overriddenDeviceSettings.set(setting, value);

    callback(removed);
    this.dispatchEventToListeners(WI.DeviceSettingsManager.Event.SettingChanged);
}
```
might even wanna make a utility for `Map.prototype.toggle`

Reply:
This is a good idea; nice and cleaner. I updated the logic with your example. Instead of `remove`, I opted for `apply` boolean to signal when we want to _apply_ the override and on the reverse we remove it.